bits.go

```diff
@@ -0,0 +1,177 @@
+package iabtcf
+
+import (
+	"fmt"
+	"time"
+)
+
+// //////////////////////////////////////////////////
+// bits
+
+// Bits represents a bitset with some helpers to read int, bool, string and time fields
+//
+// Bits are stored in a byte slice
+// First byte will store the first 8 bits, second byte the next 8 bits, and so on
+//
+// note: the last byte may contain less than 8 bits. Those bits are left aligned.
+type Bits []byte
+
+// HasBit checks if the bit number is set
+//
+// note: number is not the index and it starts at 1.
+func (b Bits) HasBit(number int) bool {
+	return b.ReadBoolField(number - 1)
+}
+
+// Length returns the number of bits in the bitset
+func (b Bits) Length() int {
+	return len(b) * nbBitInByte
+}
+
+const (
+	nbBitInByte  = 8
+	lastBitIndex = nbBitInByte - 1
+)
+
+var (
+	bitMasks = [nbBitInByte]byte{
+		1 << 7,
+		1 << 6,
+		1 << 5,
+		1 << 4,
+		1 << 3,
+		1 << 2,
+		1 << 1,
+		1,
+	}
+)
+
+// ReadInt64Field reads an int64 field of nbBits bits starting at offset
+//
+// note: if offset is negative, the result will be zero
+// note: if offset + nbBits is out of bound, the result will be the same if we were adding trailing zeros
+// example: 00101 > read with offset 2 and nbBits 5 > equivalent to reading 10100 = 20
+func (b Bits) ReadInt64Field(offset, nbBits int) int64 {
+	if offset < 0 {
+		return 0
+	}
+	var result int64
+	byteIndex := offset / nbBitInByte
+	if byteIndex >= len(b) {
+		return result
+	}
+	bitIndex := offset % nbBitInByte
+	for i := 0; i < nbBits; i++ {
+		mask := bitMasks[bitIndex]
+		if b[byteIndex]&mask == mask {
+			result |= 1 << (nbBits - 1 - i)
+		}
+		if bitIndex == lastBitIndex {
+			byteIndex++
+			if byteIndex >= len(b) {
+				return result
+			}
+			bitIndex = 0
+		} else {
+			bitIndex++
+		}
+	}
+	return result
+}
+
+// ReadIntField reads an int field of nbBits bits starting at offset
+func (b *Bits) ReadIntField(offset, nbBits int) int {
+	return int(b.ReadInt64Field(offset, nbBits))
+}
+
+const (
+	TimeNbBits = 36
```

Comment:
Does this need to be exported?  Also similar items below.

Reply:
✔️ Done